### ISSUE TITLE
TODOs for C++ GraphIndex

### ISSUE THREAD

**Issue by jermainewang** [dmlc/dgl#66]
**Implementations**
GraphIndex related:
- [x] [Graph operators](https://github.com/jermainewang/dgl/blob/cpp/include/dgl/graph_op.h): Currently, only `DisjointUnion` has been implemented.
~~- [ ] Graph traversal: A topological traversal implementation in C++ that is used by tree-lstm example.~~ (Move to next PR for improving TreeLSTM)

BatchedGraph related [batch.py](https://github.com/jermainewang/dgl/blob/cpp/python/dgl/batch.py):
- [x] Basic implementation: batch, unbatch.
~~- [ ] Need to finish other APIs: split, slice/update, readout.~~ (Move to next PR for improving DGMG)

Subgraph related: (@zheng-da )
Discussion thread #67 
- [x] For graph index, currently only node subgraph has been implemented in C++ side.
~~On python side, we need to rewrite the DGLSubgraph class to support: shared-write on node/edge features.~~(move to next PR)

**Necessary optimizations**
- [x] Zero-copy between `dgl.ndarray` and `torch.tensor`.
- [x] Fused implementation of `send_and_recv`.
- [x] Lowering degree bucketing to C++.
- [x] Benchmark the throughput of GraphIndex

**Project managements**
- [x] New install script to also compile the cpp files.
- [x] New jenkins script for CI.
- [x] I suggest we setup dockerfile for our project to ease the above jenkins setup. Could borrow from TVM again.
- [x] Setup lint check for C++ files. Borrow from dmlc-core.

**Others**
- [x] Remove batchable=False mode and related tests/codes.
- [x] Remove anonymous representation.
- [x] Double check all the models. Benchmark their speed and accuracy under the new change. (TBD: LineGraph) (DGMG and JTNN will be included in the next PR)

The goal is to make sure that once the `cpp` branch is merged, all the tests and examples are still working. Therefore, the BatchedGraph and Subgraph utilities can be incomplete. They only need to serve the current examples.
